### PR TITLE
add universal Connect func for unix,tcp and tls connection

### PR DESCRIPTION
**Pull request by vtolstov**
Signed-off-by: Vasiliy Tolstov <v.tolstov@unistack.org>

### CODE REVIEW

**Comment by vtolstov** [comment]
this is universal function to able to connect to tcp, unix and tls (via provided tls.Config), needed for https://github.com/eBay/go-ovn/pull/42

File: client.go
Comment (hzhou8) [inline client.go]: Better to rename parameters with more clear meaning:
network => protocol
config => tlsConfig

The interface name Dial() is also confusing with the Connect() interface. The names should clearly suggest what they are doing and we should make them easy for users to know which one to use. I am thinking about consolidate the interfaces into one: Connect(connMethod string, tlsConfig *tls.Config)
"connMethod" should be in the format specified by man page of ovsdb(7), section CONNECTION METHODS. For now we support only:

ssl:<host>:<port>
tcp:<host>:<port>
unix:<file>

In the future we can support more formats that are specified in ovsdb(7), such as <method1>,<method2>,...,<methodN> for clustered DB.

All the other interface should be removed (or at least documented as Deprecated). What do you think? @vtolstov @softwarejl

**Review by hzhou8 — CHANGES_REQUESTED** [review]

**Comment by vtolstov** [comment]
i'm cleanup old methods and add only one single Connect method.
I think you @hzhou8 need to tag release before this merged, and merge this after that.
Because now we use go modules and easy can use different releases of libovsdb.

File: client.go
Comment (hzhou8) [inline client.go]: As commented earlier, I think it may be better just:
Connect(connMethod string, tlsConfig *tls.Config)
"connMethod" should be in the format specified by man page of ovsdb(7), section CONNECTION METHODS.

**Review by hzhou8 — CHANGES_REQUESTED** [review]

**Comment by hzhou8** [comment]
> I think you @hzhou8 need to tag release before this merged, and merge this after that.
Because now we use go modules and easy can use different releases of libovsdb.

There seems no point to do release "before" merging this, because the old interface was already changed even before this patch. Since the new interface has been there only for 2 weeks, it is unlikely that someone other than us is already using it, before you change it again :)

**Comment by hzhou8** [comment]
btw, CI is failing

**Comment by vtolstov** [comment]
@hzhou8 now all fine as i see, please recheck

**Comment by hzhou8** [comment]
@vtolstov thanks for fixing CI. I think you missed my comment:

As commented earlier, I think it may be better just:
Connect(connMethod string, tlsConfig *tls.Config)
"connMethod" should be in the format specified by man page of ovsdb(7), section CONNECTION METHODS.

**Comment by vtolstov** [comment]
> @vtolstov thanks for fixing CI. I think you missed my comment:
> 
> As commented earlier, I think it may be better just:
> Connect(connMethod string, tlsConfig *tls.Config)
> "connMethod" should be in the format specified by man page of ovsdb(7), section CONNECTION METHODS.

By i need also to specific endpoint. So We can use current pr, or modify Connect to something like 
Connect(addr string, tlsConfig *tls.Config) addr combined con method and ip address with port (or unix socket)
@hzhou8 what is the preferred way?

**Comment by vtolstov** [comment]
@hzhou8 @softwarejl can you look?

File: client.go
Comment (softwarejl) [inline client.go:5]: could we have standare for import package here:
3 section:
1st section: default go package
2nd section: ebay dependancy.
3rd section: go-ovn dependancy.

Comment (vtolstov) [reply client.go:5]: i can't because this is goimports and gofmt it always sort standard and after it all other with alphabetical order

**Review by softwarejl — APPROVED** [review]

**Comment by vtolstov** [comment]
@hzhou8 i need you accept too...

File: client.go
Comment (hzhou8) [inline client.go:61]: Thanks for the update. However, OVSDB connect method format is not URL format. I.e. it is something like tcp:10.0.0.1:6641 instead of tcp://10.0.0.1:6641

File: client.go
Comment (hzhou8) [inline client.go]: I think it is better to add protocol here, i.e. "tcp:127.0.0.1:6640"

File: client.go
Comment (hzhou8) [inline client.go]: I don't think we need 2 default addresses. The default can be either TCP or Unix. We should follow C/Python ovsdb client patterns. There were no use case that user specifies protocol but omits the actual address.

**Review by hzhou8 — CHANGES_REQUESTED** [review]

File: client.go
Comment (vtolstov) [reply client.go:61]: i handle this, check https://play.golang.org/p/7gdThaOe7-E

File: client.go
Comment (vtolstov) [reply client.go]: ok, but i'm prefer to hide this. and use variables internally.
unix path is the default, but can vary in various distros.

File: client.go
Comment (vtolstov) [reply client.go]: i create this variables for end-user, so if you can't connect you can see this variables and check own configuration of ovn.

Comment (vtolstov) [reply client.go]: i'm un-export this variables..

File: client.go
Comment (hzhou8) [reply client.go:61]: OK, so it handles both. Thanks for explain. I was mislead by the test cases where "tcp://..." is used always. Please update the test cases to use "tcp:..." to avoid confusion, and please don't forget to update example/play_with_ovs.go.

**Review by hzhou8 — CHANGES_REQUESTED** [review]

**Comment by vtolstov** [comment]
@hzhou8 all comments addressed now, please check

**Review by hzhou8 — APPROVED** [review]

**Pull request merged**
